### PR TITLE
Added support for querying by tags

### LINKED ISSUE #1456: Package Tags
Package tags have been proposed as a way to organize packages into groups.

@tgamblin 

> w.r.t. organizing them, I was pondering adding tags (ala gmail) to the packages. It would be easy to implement and would give us a more categorical high level view of the packages themselves. What do people think of that? 
> 
> We'd probably want either another command or another argument for spack list to show this. I was thinking it might be nice if, say, "spack tags" showed total package count, available tags and number of packages with each tag (likely not distinct sets). It would also allow listing the packages with one or more tags. Thoughts? I like this better than strict hierarchical layouts used by things like Gentoo and macports.

@davydden 

> I think tags is a very good idea which would help a lot to organise packages (such as LinearAlgebra, PartialDifferentialEquations). One would need to find a good compromise in "granularity", but that should work. A github issue to discuss and agree with the list of tags which fits the current packages ought to do it.
> 
> Obviously a single package can have several tags attached. One should probably maintain the list of tags somewhere with their description and then force packages to use only those tags.

The purpose of this issue is to assemble input on a set of tags we will use canonically in Spack.  Current suggested tags are:

`
LinearAlgebra: <Any blas/lapack implementation>, eigen, py-numpy
PartialDifferentialEquations: eigen, py-scipy
`

I would suggest some others:

```
Numerical:  scipy,octave
Climate: modele, icebin
Cryosphere: icebin, pism
MolecularDynamics
...other domain-specific stuff...
DomainSpecific: (would subsume Climate, MolecularDynamics, etc)
Storage: netcdf, hdf5, mariadb, etc.
Database: mariadb
GUI: qt,gnome, etc.
Language: python, perl, r, ghc, ruby, etc.
Python,R,Perl, Haskell, Ruby, etc.
Cxx: boost
Compiler: (includes glibc as well)
Utilities: git,curl,elfutils
Typesetting: tex
SystemLibs: openssl, glibc, dbus
Supercomputer: mpi,openmp,slurm
```

Now that I'm thinking this through.... we want not just tags, but a lattice built out of tags and packages.  In layman's terms, this means that Tags can have Tags on them as well.  So... the `Cryosphere` tag would have the `Climate` tag on it.  Anything listing of `Climate` packages would also list `Cryosphere` packages (unless you're looking for packages that explicitly list `Climate`, i.e. they are not transitive).  When you query the lattice, the results should include both tags and packages (unless the user asked for only one type).

Other conventions: package names are always lower case (any violations of this convention in current Spack are subject to change).  Tag names are always upper camel case.

Also, let's make tags always singular.  No plurals.

### LINKED ISSUE #1526: add tags to packages
subj. I would say for starters, one-to-many (one package many tags) without hierarchies of tags is perfectly enough to help users better orient in packages.

As a CLI i think it's enough to have:

```
spack tags                   (show all tags with their short description, two column print out)
spack list --tag=<var>       (show all packages from a tag <var>)
spack find --tag=<var>       (show all installed packages from a tag <var>)
```

Each package declare tags via:

```
tags = one, two, three
```

with some (manual) checks that `one, two, three` are contained in the list of tags stored in `repo.yaml` as a pairs of strings `name - description`.

in reference to the discussion  https://groups.google.com/forum/?fromgroups#!topic/spack/zGs-0RKz3_U

### LINKED ISSUE #4781: Use of "tags" attribute in packages
I noticed that a few packages lately introduced the attribute `tags`, e.g. :
```python
class Cloverleaf3d(MakefilePackage):
    """Proxy Application. CloverLeaf3D is 3D version of the
       CloverLeaf mini-app. CloverLeaf is a mini-app that solves
       the compressible Euler equations on a Cartesian grid,
       using an explicit, second-order accurate method.
    """

    homepage = "http://uk-mac.github.io/CloverLeaf3D/"
    url      = "http://mantevo.org/downloads/releaseTarballs/miniapps/CloverLeaf3D/CloverLeaf3D-1.0.tar.gz"

    tags = ['proxy-app']
    ...
```
Other packages go even further, and they have a last line in their docstring providing redundant information:
```python
class Aspa(MakefilePackage):
    """A fundamental premise in ExMatEx is that scale-bridging performed in
        heterogeneous MPMD materials science simulations will place important
        demands upon the exascale ecosystem that need to be identified and
        quantified.
        tags = proxy-app
    """
    tags = ['proxy-app']
    ...
```
In #4774 `tags` is only in the docstring (no attribute), and it uses a colon as a separator.

I would like to start a discussion on this, because I think we should decide how to tag packages before introducing multiple inconsistent ways of doing it. Also, in case none of the above is tied to a query mechanism, I think we should also decide which use we want to make of this information (e.g. where it can be used to filter a query, and what should be the result).

### PR DESCRIPTION
**Pull request by alalazo**
Fixes #4781 
Fixes #1456
Fixes #1526

The querying commands `spack list`, `spack find` and `spack info` have been modified to support querying by tags. Tests have been added to check that the feature is working correctly under what I think will be the most frequent use cases.

##### spack list
```console
$ spack list --tags proxy-app
==> 8 packages.
aspa  cloverleaf3d  hpccg  miniaero  miniamr  minife  pathfinder  tealeaf
```

##### spack find
```console
$ spack find
==> 70 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
autoconf@2.69  cloverleaf3d@1.0  flex@2.6.1        hdf5@1.10.1      libevent@2.0.21      libtool@2.4.6  mpich@3.2             netlib-scalapack@2.0.2  openmpi@2.1.1      scotch@6.0.4        voropp@0.4.6
automake@1.15  cmake@3.8.0       flex@2.6.1        hdf5@1.10.1      libpciaccess@0.13.4  libxc@3.0.0    mumps@5.1.1           netlib-scalapack@2.0.2  openssl@1.0.2k     szip@2.1            xz@5.2.3
bison@3.0.4    cmake@3.8.1       gettext@0.19.8.1  help2man@1.47.4  libpciaccess@0.13.5  libxml2@2.9.4  ncurses@6.0           netlib-scalapack@2.0.2  parmetis@4.0.3     szip@2.1.1          zlib@1.2.11
boost@1.63.0   fftw@3.3.6-pl2    gflags@2.1.2      hwloc@1.11.6     libsigsegv@2.10      m4@1.4.18      netcdf@4.4.1.1        openblas@0.2.19         pkg-config@0.29.2  tar@1.29
bzip2@1.0.6    fftw@3.3.6-pl2    glog@0.3.4        hwloc@1.11.7     libsigsegv@2.11      metis@5.1.0    netcdf-fortran@4.4.4  openmpi@2.1.0           qhull@2015.2       util-macros@1.19.1

-- linux-ubuntu14-x86_64 / gcc@6.3.0 ----------------------------
autoconf@2.69  boost@1.63.0  cmake@3.8.1              folly@2017.06.05.00  glog@0.3.4       libsigsegv@2.11  m4@1.4.18    openssl@1.0.2k     zlib@1.2.11
automake@1.15  bzip2@1.0.6   double-conversion@2.0.1  gflags@2.1.2         libevent@2.0.21  libtool@2.4.6    ncurses@6.0  pkg-config@0.29.2

$ spack find --tags proxy-app
==> 1 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
cloverleaf3d@1.0
```

##### spack info
```console
$ spack info cloverleaf3d
MakefilePackage:    cloverleaf3d

Description:
    Proxy Application. CloverLeaf3D is 3D version of the CloverLeaf mini-
    app. CloverLeaf is a mini-app that solves the compressible Euler
    equations on a Cartesian grid, using an explicit, second-order accurate
    method.

Homepage:           http://uk-mac.github.io/CloverLeaf3D/

Safe versions:  
    1.0    http://mantevo.org/downloads/releaseTarballs/miniapps/CloverLeaf3D/CloverLeaf3D-1.0.tar.gz

Variants:
    Name [Default]    Allowed values    Description


    openacc [off]     True, False       Enable OpenACC Support


Installation Phases:
    edit    build    install

Build Dependencies:
    mpi

Link Dependencies:
    mpi

Run Dependencies:
    None

Virtual Packages: 
    None

Tags: 
    proxy-app

```

### CODE REVIEW

**Comment by alalazo** [comment]
> I think the next step is to come up with a common set of tags to use and a means of controlling their use. If anyone can add whatever tags they feel like, tags will become much less useful. This is even more important than consistently spelled and capitalized variants.

I have mixed feelings about "means of controlling their use". We have been dealing fine with virtual dependencies without having code that was enforcing them - just community conventions. Maybe we could do the same for tags?

**Comment by adamjstewart** [comment]
@alalazo That's true. Once a tag has been established, it will be simple to enforce people adding to the same tag. We just need to make sure that the initial tag is well-named. I'm just trying to prevent the scenario where someone adds a tag, `foo`, which secretly means "all the packages that I care about/develop", even if they aren't related.

Speaking of which, does anyone know what a `proxy-app` is???

**Comment by alalazo** [comment]
> Speaking of which, does anyone know what a proxy-app is???

Don't know, but it seems to be the same thing as a mini-app :laughing:

File: lib/spack/spack/cmd/common/arguments.py [hunk 1]
```diff
@@ -43,6 +53,44 @@ def add_common_arguments(parser, list_of_arguments):
...
+        specs_or_packages: list of items to be filtered
+        tags: tags to be used for filtering
+
+    Returns:
+        list of items in specs that have all the tag in tags
+    """
+
+    # If the list is empty, return early
+    if not specs_or_packages:
+        return
```
Comment (tgamblin) [inline lib/spack/spack/cmd/common/arguments.py]: should return an empty list.

File: lib/spack/spack/cmd/common/arguments.py [hunk 2]
```diff
@@ -43,6 +53,44 @@ def add_common_arguments(parser, list_of_arguments):
...
+
+    # Else, try to have a uniform access to packages for both list
+    # of specs and list of packages
+    try:
+        # This statement is needed to check if I am dealing
+        # with a list of specs
```
Comment (tgamblin) [inline lib/spack/spack/cmd/common/arguments.py]: why not just use `isinstance(element, spack.spec.Spec)`?  Seems like the only special case.  Also what if a package for some reason has a `package` attribute?  A package author could add one (don't want to restrict that).

File: lib/spack/spack/cmd/common/arguments.py [hunk 3]
```diff
@@ -43,6 +53,44 @@ def add_common_arguments(parser, list_of_arguments):
...
+        specs_or_packages[0].package
+        package = lambda x: x.package
+    except AttributeError:
+        package = lambda x: x
```
Comment (tgamblin) [inline lib/spack/spack/cmd/common/arguments.py]: I'd just convert the whole try/catch above to this:

```python
packages = [s.package if isinstance(s, Spec) else s
            for s in specs_or_packages]
```

Then work on the `packages` list below rather than using a lambda all the time.  Also this allows you to filter hybrid lists.

File: lib/spack/spack/cmd/common/arguments.py [hunk 4]
```diff
@@ -43,6 +53,44 @@ def add_common_arguments(parser, list_of_arguments):
...
+
+    # Restrict to elements that have the 'tags' attribute defined
+    has_tags = lambda x: hasattr(package(x), 'tags')
+    specs_or_packages = [x for x in specs_or_packages if has_tags(x)]
+
+    # Filter by required tags
+    for tag in tags:
+        specs_or_packages = [
+            x for x in specs_or_packages if tag in package(x).tags
+        ]
```
Comment (tgamblin) [inline lib/spack/spack/cmd/common/arguments.py]: I think this is more concise and more efficient (using suggested `packages` name from above):

```python
return [p for p in packages
        if any(t in tags for t in getattr(p, 'tags', []))]
```

You can replace lines 81-91 with that.

File: lib/spack/spack/cmd/info.py [hunk 5]
```diff
@@ -191,6 +194,16 @@ def print_text_info(pkg):
     else:
         print("    None")
 
+    print()
+    print("Tags: ")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/info.py]: Let's move this to the top, maybe before or after the homepage?  I think we should put more accessible information at the top of `info` and more esoteric stuff (dependencies) at the end.

File: lib/spack/spack/cmd/common/arguments.py (hunk 2 above)
Comment (alalazo) [inline lib/spack/spack/cmd/common/arguments.py]: I did this way to avoid an import, but I have no strong preferences. I'll change the `try` statement.

File: lib/spack/spack/cmd/common/arguments.py (hunk 3 above)
Comment (alalazo) [inline lib/spack/spack/cmd/common/arguments.py]: I see a problem with that. Right now we are returning a list of elements of the same type of what went in. After this modification we will return always a list of packages (which is wrong for `spack find`).

File: lib/spack/spack/cmd/common/arguments.py (hunk 4 above)
Comment (alalazo) [inline lib/spack/spack/cmd/common/arguments.py]: Concerning the logic here: do you think it should be `any` or `all`? Currently I implemented `all` but was not sure what could be better in case of multiple tags like:
```console
spack find --tags foo --tags bar
```

**Comment by alalazo** [comment]
Just to have some reference measurements, on my laptop `spack find` is 2x-3x slower:
```console
$ time spack find
==> 136 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
atk@2.20.0        curl@7.54.0        gdk-pixbuf@2.31.2             hdf5@1.10.1       libaio@0.3.110        libsigsegv@2.11    lua@5.3.4        netcdf-fortran@4.4.4    perl@5.24.1           szip@2.1.1
atk@2.20.0        expat@2.2.0        gettext@0.19.8.1              help2man@1.47.4   libedit@3.1-20170329  libtiff@4.0.8      lz4@1.7.5        netlib-scalapack@2.0.2  perl-xml-parser@2.44  tar@1.29
autoconf@2.69     fftw@3.3.6-pl2     gflags@2.1.2                  hwloc@1.11.6      libevent@2.0.21       libtool@2.4.6      m4@1.4.18        netlib-scalapack@2.0.2  pixman@0.34.0         util-macros@1.19.1
automake@1.15     fftw@3.3.6-pl2     glib@2.53.1                   hwloc@1.11.7      libffi@3.2.1          libx11@1.6.5       mariadb@10.1.23  netlib-scalapack@2.0.2  pkg-config@0.29.2     util-macros@1.19.1
bison@3.0.4       flex@2.6.1         glog@0.3.4                    hypre@2.11.2      libgcrypt@1.7.6       libxau@1.0.8       metis@5.1.0      openblas@0.2.19         python@2.7.13         voropp@0.4.6
boost@1.63.0      flex@2.6.1         gobject-introspection@1.49.2  icu4c@58.2        libgpg-error@1.27     libxc@3.0.0        metis@5.1.0      openmpi@2.1.0           qhull@2015.2          xcb-proto@1.12
bzip2@1.0.6       flex@2.6.4         gobject-introspection@1.49.2  inputproto@2.3.2  libpciaccess@0.13.4   libxcb@1.12        mfem@3.3         openmpi@2.1.1           readline@7.0          xextproto@7.3.0
cairo@1.14.8      font-util@1.3.1    gperf@3.0.4                   intltool@0.51.0   libpciaccess@0.13.5   libxdmcp@1.1.2     mpich@3.2        openssl@1.0.2k          renderproto@0.11.1    xproto@7.0.31
cairo@1.14.8      fontconfig@2.12.3  gtkplus@2.24.31               jemalloc@4.5.0    libpng@1.6.29         libxext@1.3.3      mumps@5.1.1      pango@1.40.3            scotch@6.0.4          xtrans@1.3.5
cloverleaf3d@1.0  freetype@2.7.1     harfbuzz@1.4.6                jpeg@9b           libpthread-stubs@0.4  libxft@2.3.2       munge@0.5.11     pango@1.40.3            shared-mime-info@1.8  xz@5.2.3
cmake@3.8.0       gdbm@1.13          harfbuzz@1.4.6                json-c@0.12.1     libsigsegv@2.10       libxml2@2.9.4      ncurses@6.0      parmetis@4.0.3          sqlite@3.18.0         zlib@1.2.11
cmake@3.8.1       gdk-pixbuf@2.31.2  hdf5@1.10.1                   kbproto@1.0.7     libsigsegv@2.11       libxrender@0.9.10  netcdf@4.4.1.1   pcre@8.40               szip@2.1

-- linux-ubuntu14-x86_64 / gcc@6.3.0 ----------------------------
autoconf@2.69  boost@1.63.0  cmake@3.8.1              folly@2017.06.05.00  glog@0.3.4       libsigsegv@2.11  m4@1.4.18    openssl@1.0.2k     zlib@1.2.11
automake@1.15  bzip2@1.0.6   double-conversion@2.0.1  gflags@2.1.2         libevent@2.0.21  libtool@2.4.6    ncurses@6.0  pkg-config@0.29.2

real	0m0.331s
user	0m0.296s
sys	0m0.024s

$ time spack find --tags=proxy-app
==> 1 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
cloverleaf3d@1.0

real	0m0.821s
user	0m0.784s
sys	0m0.028s
```
while `spack list` is one order of magnitude slower with tags:
```console
$ time spack list
==> 1659 packages.
...
real	0m0.341s
user	0m0.232s
sys	0m0.056s

$ time spack list --tags=proxy-app
==> 21 packages.
amr-exp-parabolic  aspa  clamr  cloverleaf3d  cns-nospec  cohmm  cosp2  ebms  hpccg  miniaero  miniamr  minife  nekbone  nut  openmc  pathfinder  pennant  simplemoc  smc  tealeaf  xsbench

real	0m3.969s
user	0m3.816s
sys	0m0.140s
```

**Comment by alalazo** [comment]
So, after having added a cache for tags:
```console
$ spack purge -m

$ time spack list --tags proxy-app
==> 21 packages.
amr-exp-parabolic  aspa  clamr  cloverleaf3d  cns-nospec  cohmm  cosp2  ebms  hpccg  miniaero  miniamr  minife  nekbone  nut  openmc  pathfinder  pennant  simplemoc  smc  tealeaf  xsbench

real	0m3.772s
user	0m3.632s
sys	0m0.116s

$ time spack list --tags proxy-app
==> 21 packages.
amr-exp-parabolic  aspa  clamr  cloverleaf3d  cns-nospec  cohmm  cosp2  ebms  hpccg  miniaero  miniamr  minife  nekbone  nut  openmc  pathfinder  pennant  simplemoc  smc  tealeaf  xsbench

real	0m0.282s
user	0m0.248s
sys	0m0.020s
```
Similarly for `spack find`:
```console
$ spack purge -m

$ time spack find --tags proxy-app
==> 1 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
cloverleaf3d@1.0

real	0m3.835s
user	0m3.624s
sys	0m0.188s

$ time spack find --tags proxy-app
==> 1 installed packages.
-- linux-ubuntu14-x86_64 / gcc@4.8 ------------------------------
cloverleaf3d@1.0

real	0m0.332s
user	0m0.280s
sys	0m0.040s

```

File: lib/spack/spack/repository.py [hunk 6]
```diff
@@ -93,6 +101,242 @@ def __getattr__(self, name):
...
+            if pkg_name in pkg_list:
+                pkg_list.remove(pkg_name)
+
+        # Add it again under the appropriate tags
+        for tag in getattr(package, 'tags', []):
+            self._tag_dict[tag].append(package.name)
+
+
+@llnl.util.lang.memoized
+def make_provider_index_cache(packages_path, namespace):
```
Comment (alalazo) [inline lib/spack/spack/repository.py]: @tgamblin This function shares most code with `make_tag_index_cache`. We can refactor them together in a later PR, if you don't think it's a big issue. The idea is that we can take a better look at where `ProviderIndex` is used, and maybe change it to be a `Mapping` too. Then refactor these two functions when the interfaces will be more homogeneous.

**Comment by alalazo** [comment]
@tgamblin I added the cache here and rebased, so I think this PR is ready for another review.  I also had a look at making tests consistent with what was done in #4868, but as it requires some work I would prefer to do it in another PR. The issue is that in `spack info` we use `print` directly, but we don't pass the correct `file` argument to retrieve the output.

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
I think this is also ready to be reviewed again. The tests issues will probably be solved by #5084
